ENT-201: Add support for course run key in catalog contains endpoint

zubair-arbi opened the pull request:
ENT-201

@saleem-latif @asadiqbal08 @mattdrayer @clintonb @cpennington 
Update the endpoint `contains` of `CatalogViewSet` to accept a new optional parameter `course_run_id`. This new parameter will be used for validating if the course run keys of the format `org/course/run` or `course-v1:org+course+run` exists in a catalog.

For example in `Ecom` we can use this endpoint like this:
```python
>>> course_ids = 'edX+TEST_01'
>>> course_run_ids = 'org/course/run,course-v1:edX+DemoX+Demo_Course'
>>> catalog_id = 1
>>> response = site.siteconfiguration.course_catalog_api_client.catalogs(catalog_id).contains.get(
    course_id=course_ids,
    course_run_id=course_run_ids,
)
>>> response
{u'courses': {u'edX+TEST_01': True, u'org/course/run': False, u'course-v1:edX+DemoX+Demo_Course': True}}
```

**Related PR:** [ENT-211 add aggregate course key](https://github.com/edx/opaque-keys/pull/87)

zubair-arbi commented:
@clintonb Please review this PR and let me know your thoughts on whether we still need the PR [ENT-211 add aggregate course key](https://github.com/edx/opaque-keys/pull/87)?
One possibility I can think of for the aggregate course key is for validating the input key provided to the catalog contains endpoint.

Our base requirement was to validate if a course run in the basket is available for a catalog.

clintonb commented on course_discovery/apps/api/v1/tests/test_views/test_catalogs.py:
```diff
@@ -153,14 +153,28 @@ def test_courses(self):
         assert response.data['results'] == self.serialize_catalog_course(courses, many=True)
 
     def test_contains(self):
-        """ Verify the endpoint returns a filtered list of courses contained in the catalog. """
```
This change is not necessary.

clintonb replied:
Also, split this into two separate test cases, or use `ddt` to achieve a similar result.

clintonb commented on course_discovery/apps/api/v1/views/catalogs.py:
```diff
@@ -113,12 +113,26 @@ def contains(self, request, id=None):  # pylint: disable=redefined-builtin,unuse
+
+        if course_run_ids:
+            course_run_ids = course_run_ids.split(',')
+            course_runs = CourseRun.search(catalog.query).filter(key__in=course_run_ids).values_list('key', flat=True)
```
Move this to a method on the `Catalog` model, similar to `Catalog.contains()`.

zubair-arbi commented:
@clintonb Please have another look

clintonb commented on course_discovery/apps/catalogs/tests/test_models.py:
```diff
@@ -39,6 +39,17 @@ def test_contains(self):
+        """ Verify the method returns a mapping of course run IDs to booleans. """
+        partner = PartnerFactory()
+        course_run = CourseRunFactory(course__partner=partner, course=self.course)
+        uncontained_course = CourseFactory(key='d/e/f', title='ABDEF')
```
This line is not necessary.

clintonb commented on course_discovery/apps/catalogs/tests/test_models.py:
```diff
@@ -39,6 +39,17 @@ def test_contains(self):
+    def test_contains_course_runs(self):
+        """ Verify the method returns a mapping of course run IDs to booleans. """
+        partner = PartnerFactory()
+        course_run = CourseRunFactory(course__partner=partner, course=self.course)
```
`course__partner` is never used since you pass `course`.

clintonb commented on course_discovery/apps/catalogs/tests/test_models.py:
```diff
@@ -39,6 +39,17 @@ def test_contains(self):
+        partner = PartnerFactory()
+        course_run = CourseRunFactory(course__partner=partner, course=self.course)
+        uncontained_course = CourseFactory(key='d/e/f', title='ABDEF')
+        uncontained_course_run = CourseRunFactory(course__partner=partner, course=uncontained_course)
```
`uncontained_course_run = CourseRunFactory(title_override='ABD')`

zubair-arbi commented:
@clintonb thanks for the valuable feedback. I have addressed your suggestions.
@mattdrayer please give review.
These changes should give live before its related changes in [Ecom - use catalog contains endpoint for varifying the course seats against a course catalog range](https://github.com/edx/ecommerce/pull/1195).

mattdrayer commented:
Thanks for taking care of this @zubair-arbi  -- LGTM 👍
